# Benchmark
[Text Embedding](https://python.langchain.com/docs/integrations/text_embedding/)

In [21]:
import markdown_splitter as mdS
from langchain.vectorstores import LanceDB

docs = mdS.create_db(path='../', glob="**/*.md",split=None)

query = "How install the chatbots"

100%|██████████| 26/26 [00:00<00:00, 3309.22it/s]


# BGE  & LanceDB

In [22]:
%%time 

# Embeddings
from langchain_community.embeddings import HuggingFaceBgeEmbeddings
embeddings = HuggingFaceBgeEmbeddings(model_name="BAAI/bge-small-en", model_kwargs={"device": "cpu"}, encode_kwargs={"normalize_embeddings": True})

# Store
docsearch = LanceDB.from_documents(docs, embeddings)

CPU times: user 38 s, sys: 14.2 s, total: 52.3 s
Wall time: 16.7 s


In [23]:
%%time
for i in range(10):
    docsearch.similarity_search(query)

docsearch.similarity_search(query)

CPU times: user 950 ms, sys: 56.3 ms, total: 1.01 s
Wall time: 266 ms


[Document(page_content='* [ChatBot Llama](https://blog.streamlit.io/how-to-build-a-llama-2-chatbot/)\n* [ChatBot Llama Exemple](https://llama2.streamlit.app)  \n* [ChatBot Streamlit](https://github.com/streamlit/llm-examples/tree/main)\n* [ChatBot Streamlit Exemple](https://llm-examples.streamlit.app)  \n* [ChatGPT API](https://platform.openai.com/docs/introduction)\n* [fine-tuning Chat-GPT](https://platform.openai.com/docs/guides/fine-tuning)  \n* LangChain :\n* [Hugging Face](https://python.langchain.com/docs/integrations/platforms/huggingface)\n* [Hugging Face Tool](https://python.langchain.com/docs/integrations/tools/huggingface_tools)  \n* [Docs GitHub](https://python.langchain.com/docs/integrations/document_loaders/github)\n* [Docs Source Code](https://python.langchain.com/docs/integrations/document_loaders/source_code)  \n* [subtitles](https://python.langchain.com/docs/integrations/document_loaders/imsdb)', metadata={'vector': [-0.06987015902996063, -0.0037282879929989576, -0.00

# FastEmbed & LanceDB

In [24]:
%%time 

# Embeddings
from langchain_community.embeddings.fastembed import FastEmbedEmbeddings
embeddings = FastEmbedEmbeddings()

# Store
docsearch = LanceDB.from_documents(docs, embeddings)

Fetching 9 files:   0%|          | 0/9 [00:00<?, ?it/s]

CPU times: user 1min 9s, sys: 2.01 s, total: 1min 11s
Wall time: 21.1 s


In [25]:
%%time
for i in range(10):
    docsearch.similarity_search(query)

docsearch.similarity_search(query)

CPU times: user 1.17 s, sys: 35.1 ms, total: 1.2 s
Wall time: 354 ms


[Document(page_content='* [ChatBot Llama](https://blog.streamlit.io/how-to-build-a-llama-2-chatbot/)\n* [ChatBot Llama Exemple](https://llama2.streamlit.app)  \n* [ChatBot Streamlit](https://github.com/streamlit/llm-examples/tree/main)\n* [ChatBot Streamlit Exemple](https://llm-examples.streamlit.app)  \n* [ChatGPT API](https://platform.openai.com/docs/introduction)\n* [fine-tuning Chat-GPT](https://platform.openai.com/docs/guides/fine-tuning)  \n* LangChain :\n* [Hugging Face](https://python.langchain.com/docs/integrations/platforms/huggingface)\n* [Hugging Face Tool](https://python.langchain.com/docs/integrations/tools/huggingface_tools)  \n* [Docs GitHub](https://python.langchain.com/docs/integrations/document_loaders/github)\n* [Docs Source Code](https://python.langchain.com/docs/integrations/document_loaders/source_code)  \n* [subtitles](https://python.langchain.com/docs/integrations/document_loaders/imsdb)', metadata={'vector': [-0.10050611197948456, -0.0003868783242069185, -0.00

# GPT4All & LanceDB

In [26]:
%%time 

# Embeddings
from langchain_community.embeddings import GPT4AllEmbeddings
embeddings = GPT4AllEmbeddings()

# Store
docsearch = LanceDB.from_documents(docs, embeddings)

bert_load_from_file: gguf version     = 2
bert_load_from_file: gguf alignment   = 32
bert_load_from_file: gguf data offset = 695552
bert_load_from_file: model name           = BERT
bert_load_from_file: model architecture   = bert
bert_load_from_file: model file type      = 1
bert_load_from_file: bert tokenizer vocab = 30522
CPU times: user 21.8 s, sys: 1.03 s, total: 22.9 s
Wall time: 6.01 s


In [27]:
%%time
for i in range(10):
    docsearch.similarity_search(query)

docsearch.similarity_search(query)

CPU times: user 1.15 s, sys: 27.8 ms, total: 1.18 s
Wall time: 343 ms


[Document(page_content="The chatbots are in their respective folders with an application launcher `autorun.sh` to install each of them without specific knowledge.  \nOn Mac, open the terminal and type:\n```shell\ncd\n```\nDrag the **`folder`** containing the file `autorun.sh`, then press the Enter key (↩︎).  \n_If you have done it correctly, the **`~`** between your machine's name (`name@MacBook-Pro-of-Name`) and the **`%`** sign should display the name of the `folder` instead._  \nExecute the following line of code by pressing the Enter key (↩︎):\n```shell\nsh autorun.sh\n```\nWait a moment, the model should open in your default web browser.", metadata={'vector': [-0.00478768115863204, -0.04618072137236595, 0.02846730872988701, -0.03609825298190117, 0.03550594300031662, -0.03747408464550972, -0.037277232855558395, 0.04082599654793739, 0.046806637197732925, 0.020041460171341896, 0.04236359894275665, -0.03443154692649841, 0.02185412310063839, 0.03714746609330177, 0.09091991931200027, 0.

# Hugging Face & LanceDB

In [28]:
%%time 

# Embeddings
from langchain_community.embeddings import HuggingFaceEmbeddings
embeddings = HuggingFaceEmbeddings()

# Store
docsearch = LanceDB.from_documents(docs, embeddings)

CPU times: user 1min 12s, sys: 14.8 s, total: 1min 27s
Wall time: 26.5 s


In [29]:
%%time
for i in range(10):
    docsearch.similarity_search(query)

docsearch.similarity_search(query)

CPU times: user 2.07 s, sys: 74.7 ms, total: 2.15 s
Wall time: 564 ms


[Document(page_content="The chatbots are in their respective folders with an application launcher `autorun.sh` to install each of them without specific knowledge.  \nOn Mac, open the terminal and type:\n```shell\ncd\n```\nDrag the **`folder`** containing the file `autorun.sh`, then press the Enter key (↩︎).  \n_If you have done it correctly, the **`~`** between your machine's name (`name@MacBook-Pro-of-Name`) and the **`%`** sign should display the name of the `folder` instead._  \nExecute the following line of code by pressing the Enter key (↩︎):\n```shell\nsh autorun.sh\n```\nWait a moment, the model should open in your default web browser.", metadata={'vector': [0.04661260545253754, 0.011096074245870113, 0.004561483394354582, -0.013605601154267788, -0.01694190874695778, 0.04158993810415268, 0.0781480222940445, -0.031065430492162704, 0.1168658435344696, -0.05304231494665146, -0.080237478017807, 0.005059852730482817, -0.021440759301185608, 0.0430215448141098, -0.01688515953719616, -0.

# ModelScope & LanceDB

In [30]:
%%time

# Embeddings
from langchain_community.embeddings import ModelScopeEmbeddings
model_id = "damo/nlp_corom_sentence-embedding_english-base"
embeddings = ModelScopeEmbeddings(model_id=model_id)

# Store
docsearch = LanceDB.from_documents(docs, embeddings)

2024-03-06 19:19:05,923 - modelscope - WARNING - Model revision not specified, use revision: v1.0.0
2024-03-06 19:19:06,550 - modelscope - INFO - initiate model from /Users/gauthierrammault/.cache/modelscope/hub/damo/nlp_corom_sentence-embedding_english-base
2024-03-06 19:19:06,551 - modelscope - INFO - initiate model from location /Users/gauthierrammault/.cache/modelscope/hub/damo/nlp_corom_sentence-embedding_english-base.
2024-03-06 19:19:06,555 - modelscope - INFO - initialize model from /Users/gauthierrammault/.cache/modelscope/hub/damo/nlp_corom_sentence-embedding_english-base
2024-03-06 19:19:08,778 - modelscope - WARNING - No preprocessor field found in cfg.
2024-03-06 19:19:08,779 - modelscope - WARNING - No val key and type key found in preprocessor domain of configuration.json file.
2024-03-06 19:19:08,780 - modelscope - WARNING - Cannot find available config to build preprocessor at mode inference, current config: {'model_dir': '/Users/gauthierrammault/.cache/modelscope/hub/

CPU times: user 35.2 s, sys: 2.52 s, total: 37.7 s
Wall time: 14.2 s


In [31]:
%%time
for i in range(10):
    docsearch.similarity_search(query)

docsearch.similarity_search(query)

CPU times: user 1.91 s, sys: 61.5 ms, total: 1.97 s
Wall time: 509 ms


[Document(page_content="The chatbots are in their respective folders with an application launcher `autorun.sh` to install each of them without specific knowledge.  \nOn Mac, open the terminal and type:\n```shell\ncd\n```\nDrag the **`folder`** containing the file `autorun.sh`, then press the Enter key (↩︎).  \n_If you have done it correctly, the **`~`** between your machine's name (`name@MacBook-Pro-of-Name`) and the **`%`** sign should display the name of the `folder` instead._  \nExecute the following line of code by pressing the Enter key (↩︎):\n```shell\nsh autorun.sh\n```\nWait a moment, the model should open in your default web browser.", metadata={'vector': [-0.5201736092567444, 0.27423185110092163, -0.29248499870300293, 0.04063499718904495, 0.8827818036079407, 0.3274809420108795, 0.27746108174324036, 0.20120948553085327, 0.8436272740364075, 0.346975177526474, -0.9604008793830872, 1.5929116010665894, -0.7470643520355225, -0.05567793548107147, -0.3151862919330597, -0.699335277080

# OpenClip & LanceDB

In [34]:
%%time

# Embeddings
from langchain_experimental.open_clip import OpenCLIPEmbeddings
embeddings = OpenCLIPEmbeddings()

# Store
docsearch = LanceDB.from_documents(docs, embeddings)

CPU times: user 2min 37s, sys: 48.4 s, total: 3min 25s
Wall time: 1min 45s


In [35]:
%%time
for i in range(10):
    docsearch.similarity_search(query)

docsearch.similarity_search(query)

CPU times: user 17.6 s, sys: 1.94 s, total: 19.6 s
Wall time: 5.17 s


[Document(page_content='Repository to store minecraft plugins', metadata={'vector': [-0.03832295164465904, -0.005526179913431406, -0.001729077659547329, -0.001534058596007526, -4.6630179895146284e-06, 0.03187521919608116, 0.009160798974335194, 0.03353673592209816, -0.02440008521080017, -0.005990446545183659, -0.024837587028741837, 0.0034549038391560316, 0.017462633550167084, 0.025209836661815643, 0.011942978017032146, 0.008954556658864021, 0.02475087344646454, -0.0023693570401519537, -0.013977093622088432, -0.0027789794839918613, -0.01622721366584301, 0.015238866209983826, 0.027493717148900032, -0.02464817464351654, 0.007996466010808945, -0.003760037012398243, -0.006175331771373749, -0.03940018638968468, -0.021464023739099503, -0.029513951390981674, 0.013438134454190731, -0.0043294839560985565, -0.008732677437365055, -0.03100011870265007, -0.017645543441176414, 0.004231972619891167, -0.049479179084300995, -0.0001578606606926769, 0.024368222802877426, -0.038204021751880646, -0.013329388

# HuggingFaceEmbeddings & LanceDB

In [36]:
%%time

# Embeddings
from langchain_community.embeddings import HuggingFaceEmbeddings
embeddings = HuggingFaceEmbeddings()

# Store
docsearch = LanceDB.from_documents(docs, embeddings)

CPU times: user 1min 9s, sys: 13.3 s, total: 1min 23s
Wall time: 25.6 s


In [37]:
%%time
for i in range(10):
    docsearch.similarity_search(query)

docsearch.similarity_search(query)

CPU times: user 2.02 s, sys: 84.4 ms, total: 2.1 s
Wall time: 551 ms


[Document(page_content="The chatbots are in their respective folders with an application launcher `autorun.sh` to install each of them without specific knowledge.  \nOn Mac, open the terminal and type:\n```shell\ncd\n```\nDrag the **`folder`** containing the file `autorun.sh`, then press the Enter key (↩︎).  \n_If you have done it correctly, the **`~`** between your machine's name (`name@MacBook-Pro-of-Name`) and the **`%`** sign should display the name of the `folder` instead._  \nExecute the following line of code by pressing the Enter key (↩︎):\n```shell\nsh autorun.sh\n```\nWait a moment, the model should open in your default web browser.", metadata={'vector': [0.04661260545253754, 0.011096074245870113, 0.004561483394354582, -0.013605601154267788, -0.01694190874695778, 0.04158993810415268, 0.0781480222940445, -0.031065430492162704, 0.1168658435344696, -0.05304231494665146, -0.080237478017807, 0.005059852730482817, -0.021440759301185608, 0.0430215448141098, -0.01688515953719616, -0.

# SpaCy & LanceDB

In [38]:
%%time

# Embeddings
from langchain_community.embeddings.spacy_embeddings import SpacyEmbeddings
import spacy
spacy.prefer_gpu()
embeddings = SpacyEmbeddings(model_name="en_core_web_sm")

# Store
docsearch = LanceDB.from_documents(docs, embeddings)

CPU times: user 2.3 s, sys: 395 ms, total: 2.7 s
Wall time: 2.85 s


In [39]:
%%time
for i in range(10):
    docsearch.similarity_search(query)

docsearch.similarity_search(query)

CPU times: user 96.9 ms, sys: 11.1 ms, total: 108 ms
Wall time: 105 ms


[Document(page_content='This repository hosts the docker-compose file to run the api, bot and database to emulate the backend environment.', metadata={'vector': [-0.1579895317554474, -0.49220311641693115, 0.08322589099407196, 0.1420513242483139, -0.221888929605484, 0.512469470500946, 0.4462919235229492, 0.2459588199853897, -0.13236652314662933, 0.0992526113986969, 0.10619334876537323, -0.038672883063554764, -0.09302673488855362, -0.1332067847251892, -0.23593328893184662, 0.04897136613726616, -0.2950063645839691, 0.05972540006041527, -0.37382736802101135, -0.11330032348632812, -0.11022507399320602, 0.16863252222537994, -0.1665152907371521, -0.5110393166542053, 0.5459228754043579, -0.06306955963373184, 0.5483529567718506, 0.3027847111225128, 0.12319616973400116, 0.21928717195987701, 0.007682331372052431, -0.25374460220336914, 0.3451727032661438, -0.28886258602142334, 0.11067533493041992, -0.1772521287202835, 0.4388265609741211, -0.025215035304427147, -0.13632573187351227, 0.1338228881359

# Result 

| Embeddings Module     | Speed Create table (CPU) | Speed Search (CPU) | Good Answer      |
|-----------------------|--------------------------|--------------------|:----------------:|
| BGE                   |52.3 s                    |1.01 s              |🟠                |
| FastEmbed             |1min 11s                  |1.20 s              |🟠                |
| GPT4All               |22.9 s                    |1.18 s              |🟢                |
| Hugging Face          |1min 27s                  |2.15 s              |🟢                |
| ModelScope            |37.7 s                    |1.97 s              |🟢                |
| OpenClip              |                          |                    |🔴                |
| HuggingFaceEmbeddings |1min 23s                  |2.10 s              |🟢                |
| SpaCy                 |                          |                    |🔴                |